File: app/src/main/java/com/teambrella/android/ui/team/feed/FeedAdapter.java
package com.teambrella.android.ui.team.feed;

import android.content.Context;
import android.content.res.Resources;
import android.support.v7.widget.RecyclerView;
import android.text.Html;
import android.text.format.DateUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.google.gson.JsonArray;
import com.squareup.picasso.Picasso;
import com.squareup.picasso.RequestCreator;
import com.teambrella.android.R;
import com.teambrella.android.api.TeambrellaModel;
import com.teambrella.android.api.model.json.JsonWrapper;
import com.teambrella.android.api.server.TeambrellaServer;
import com.teambrella.android.data.base.IDataPager;
import com.teambrella.android.image.TeambrellaImageLoader;
import com.teambrella.android.ui.IMainDataHost;
import com.teambrella.android.ui.base.TeambrellaDataPagerAdapter;
import com.teambrella.android.ui.chat.ChatActivity;
import com.teambrella.android.ui.widget.TeambrellaAvatarsWidgets;
import com.teambrella.android.util.TeambrellaDateUtils;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Locale;
import java.util.TimeZone;

import io.reactivex.Observable;
import jp.wasabeef.picasso.transformations.CropCircleTransformation;
import jp.wasabeef.picasso.transformations.MaskTransformation;

/**
 * Feed Adapter
 */
class FeedAdapter extends TeambrellaDataPagerAdapter {


    private static SimpleDateFormat mSDF = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.US);

    private static final int VIEW_TYPE_HEADER = VIEW_TYPE_REGULAR + 1;
    private static final int VIEW_TYPE_ITEM_FEED = VIEW_TYPE_REGULAR + 2;


    private final int mTeamId;
    private final IMainDataHost mDataHost;

    FeedAdapter(IMainDataHost dataHost, IDataPager<JsonArray> pager, int teamId) {
        super(pager);
        mTeamId = teamId;
        mDataHost = dataHost;
    }


    static {
        mSDF.setTimeZone(TimeZone.getTimeZone("UTC"));
    }

    @Override
    public int getItemViewType(int position) {
        int viewType = super.getItemViewType(position);
        if (viewType == VIEW_TYPE_REGULAR) {
            viewType = position == 0 && mDataHost.isFullTeamAccess() ? VIEW_TYPE_HEADER : VIEW_TYPE_ITEM_FEED;
        }
        return viewType;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        RecyclerView.ViewHolder viewHolder = super.onCreateViewHolder(parent, viewType);
        if (viewHolder == null) {
            LayoutInflater inflater = LayoutInflater.from(parent.getContext());
            switch (viewType) {
                case VIEW_TYPE_ITEM_FEED:
                    viewHolder = new FeedItemViewHolder(inflater.inflate(R.layout.list_item_feed_claim, parent, false));
                    break;
                case VIEW_TYPE_HEADER:
                    viewHolder = new FeedHeader(inflater.inflate(R.layout.list_item_feed_header, parent, false));
                    break;

            }
        }
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        super.onBindViewHolder(holder, position);
        if (holder instanceof FeedItemViewHolder) {
            ((FeedItemViewHolder) holder).bind(new JsonWrapper(mPager.getLoadedData().get(position - (mDataHost.isFullTeamAccess() ? 1 : 0)).getAsJsonObject()));
        }
    }

    @Override
    public int getItemCount() {
        return super.getItemCount() + (mDataHost.isFullTeamAccess() ? 1 : 0);
    }

    class FeedItemViewHolder extends RecyclerView.ViewHolder {

        private Picasso picasso;
        private ImageView mIcon;
        private TextView mTitle;
        private TextView mWhen;
        private TextView mMessage;
        private TeambrellaAvatarsWidgets mAvatarWidgets;
        private TextView mUnread;
        private TextView mType;

        FeedItemViewHolder(View itemView) {
            super(itemView);
            mIcon = itemView.findViewById(R.id.icon);
            mTitle = itemView.findViewById(R.id.title);
            mWhen = itemView.findViewById(R.id.when);
            mMessage = itemView.findViewById(R.id.message);
            picasso = TeambrellaImageLoader.getInstance(itemView.getContext()).getPicasso();
            mAvatarWidgets = itemView.findViewById(R.id.avatars);
            mUnread = itemView.findViewById(R.id.unread);
            mType = itemView.findViewById(R.id.type);
        }

        void bind(JsonWrapper item) {
            int itemType = item.getInt(TeambrellaModel.ATTR_DATA_ITEM_TYPE);
            Context context = itemView.getContext();
            Resources resources = context.getResources();
            RequestCreator requestCreator = picasso.load(TeambrellaModel.getImage(TeambrellaServer.BASE_URL, item.getObject(), TeambrellaModel.ATTR_DATA_SMALL_PHOTO_OR_AVATAR));


            if (itemType == TeambrellaModel.FEED_ITEM_TEAMMATE
                    || itemType == TeambrellaModel.FEED_ITEM_TEAM_CHAT) {
                requestCreator.transform(new CropCircleTransformation());
            } else {
                requestCreator.resize(resources.getDimensionPixelSize(R.dimen.image_size_48), resources.getDimensionPixelSize(R.dimen.image_size_48))
                        .centerCrop().transform(new MaskTransformation(context, R.drawable.teammate_object_mask));
            }

            requestCreator.into(mIcon);
            mMessage.setText(Html.fromHtml(item.getString(TeambrellaModel.ATTR_DATA_TEXT)));


            switch (itemType) {
                case TeambrellaModel.FEED_ITEM_CLAIM:
                    mType.setText(R.string.claim);
                    mType.setCompoundDrawablesWithIntrinsicBounds(resources.getDrawable(R.drawable.ic_claim), null, null, null);
                    break;
                case TeambrellaModel.FEED_ITEM_TEAM_CHAT:
                    mType.setCompoundDrawablesWithIntrinsicBounds(resources.getDrawable(R.drawable.ic_discussion), null, null, null);
                    mType.setText(R.string.discussion);
                    break;
                default:
                    mType.setCompoundDrawablesWithIntrinsicBounds(resources.getDrawable(R.drawable.ic_application), null, null, null);
                    mType.setText(R.string.application);
                    break;
            }

            switch (itemType) {
                case TeambrellaModel.FEED_ITEM_TEAM_CHAT:
                    mTitle.setText(item.getString(TeambrellaModel.ATTR_DATA_CHAT_TITLE));
                    break;
                default:
                    mTitle.setText(item.getString(TeambrellaModel.ATTR_DATA_MODEL_OR_NAME));
            }

            int unreadCount = item.getInt(TeambrellaModel.ATTR_DATA_UNREAD_COUNT);

            mUnread.setText(item.getString(TeambrellaModel.ATTR_DATA_UNREAD_COUNT));
            mUnread.setVisibility(unreadCount > 0 ? View.VISIBLE : View.INVISIBLE);


            try {
                long time = TeambrellaDateUtils.getServerTime(item.getString(TeambrellaModel.ATTR_DATA_ITEM_DATE));
                long now = System.currentTimeMillis();
                mWhen.setText(DateUtils.getRelativeTimeSpanString(time, now, DateUtils.MINUTE_IN_MILLIS, DateUtils.FORMAT_ABBREV_RELATIVE));
            } catch (ParseException e) {
                e.printStackTrace();
            }


            Observable.
                    fromIterable(item.getJsonArray(TeambrellaModel.ATTR_DATA_TOP_POSTER_AVATARS))
                    .map(jsonElement -> TeambrellaServer.BASE_URL + jsonElement.getAsString())
                    .toList()
                    .subscribe(mAvatarWidgets::setAvatars);


            itemView.setOnClickListener(v -> {
                switch (itemType) {
                    case TeambrellaModel.FEED_ITEM_CLAIM:
                        ChatActivity.startClaimChat(context
                                , mTeamId
                                , item.getInt(TeambrellaModel.ATTR_DATA_ITEM_ID)
                                , item.getString(TeambrellaModel.ATTR_DATA_MODEL_OR_NAME)
                                , TeambrellaImageLoader.getImageUri(item.getString(TeambrellaModel.ATTR_DATA_SMALL_PHOTO_OR_AVATAR))
                                , item.getString(TeambrellaModel.ATTR_DATA_TOPIC_ID)
                                , mDataHost.getTeamAccessLevel());
                        break;
                    case TeambrellaModel.FEED_ITEM_TEAM_CHAT:
                        ChatActivity.startFeedChat(context
                                , item.getString(TeambrellaModel.ATTR_DATA_CHAT_TITLE)
                                , item.getString(TeambrellaModel.ATTR_DATA_TOPIC_ID)
                                , mDataHost.getTeamAccessLevel());
                        break;
                    default:
                        ChatActivity.startTeammateChat(context, mTeamId
                                , item.getString(TeambrellaModel.ATTR_DATA_ITEM_USER_ID)
                                , null
                                , TeambrellaImageLoader.getImageUri(item.getString(TeambrellaModel.ATTR_DATA_SMALL_PHOTO_OR_AVATAR))
                                , item.getString(TeambrellaModel.ATTR_DATA_TOPIC_ID)
                                , mDataHost.getTeamAccessLevel());
                        break;

                }
            });

        }
    }

    private class FeedHeader extends RecyclerView.ViewHolder {
        FeedHeader(View itemView) {
            super(itemView);
            itemView.findViewById(R.id.start_new_discussion).setOnClickListener(v -> mDataHost.startNewDiscussion());
        }
    }

}

File: app/src/main/java/com/teambrella/android/ui/team/feed/FeedAdapter.kt
package com.teambrella.android.ui.team.feed

import android.graphics.Color
import android.support.v7.widget.RecyclerView
import android.text.Html
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.google.gson.JsonArray
import com.google.gson.JsonObject
import com.teambrella.android.R
import com.teambrella.android.api.*
import com.teambrella.android.data.base.IDataPager
import com.teambrella.android.ui.IMainDataHost
import com.teambrella.android.ui.base.TeambrellaDataPagerAdapter
import com.teambrella.android.ui.chat.ChatActivity
import com.teambrella.android.ui.util.*
import com.teambrella.android.ui.widget.TeambrellaAvatarsWidgets


const val VIEW_TYPE_HEADER = TeambrellaDataPagerAdapter.VIEW_TYPE_REGULAR + 1
const val VIEW_TYPE_ITEM_FEED = TeambrellaDataPagerAdapter.VIEW_TYPE_REGULAR + 2

class KFeedAdapter(val dataHost: IMainDataHost, val teamId: Int
                   , pager: IDataPager<JsonArray>, listener: OnStartActivityListener) : TeambrellaDataPagerAdapter(pager, listener) {

    init {
        setHasStableIds(true)
    }

    override fun getItemId(position: Int) = position.toLong()

    override fun getHeadersCount() = if (dataHost.isFullTeamAccess) 1 else 0

    override fun getItemViewType(position: Int): Int {
        val viewType = super.getItemViewType(position)
        return if (viewType == VIEW_TYPE_REGULAR) {
            if (position == 0 && dataHost.isFullTeamAccess) {
                VIEW_TYPE_HEADER
            } else {
                VIEW_TYPE_ITEM_FEED
            }
        } else {
            viewType
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        var viewHolder = super.onCreateViewHolder(parent, viewType)
        if (viewHolder == null) {
            val inflater = LayoutInflater.from(parent.context)
            when (viewType) {
                VIEW_TYPE_ITEM_FEED -> viewHolder = FeedItemViewHolder(inflater.inflate(R.layout.list_item_feed_claim, parent, false))
                VIEW_TYPE_HEADER -> viewHolder = FeedHeader(inflater.inflate(R.layout.list_item_feed_header, parent, false))
            }
        }
        return viewHolder
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        super.onBindViewHolder(holder, position)
        if (holder is FeedItemViewHolder) {
            mPager.loadedData[position - headersCount]?.asJsonObject?.let {
                holder.onBind(it)
            }
        }
    }

    override fun onPagerItemChanged(item: Int) {
        super.onPagerItemChanged(item)
        //notifyItemChanged(headersCount + item) // see https://stackoverflow.com/questions/30220771/recyclerview-inconsistency-detected-invalid-item-position
        notifyDataSetChanged();
    }

    private inner class FeedItemViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val pane: View? = itemView.findViewById(R.id.pane)
        private val icon: ImageView? = itemView.findViewById(R.id.icon)
        private val title: TextView? = itemView.findViewById(R.id.title)
        private val date: TextView? = itemView.findViewById(R.id.`when`)
        private val message: TextView? = itemView.findViewById(R.id.message)
        private val avatars: TeambrellaAvatarsWidgets? = itemView.findViewById(R.id.avatars)
        private val unread: TextView? = itemView.findViewById(R.id.unread)
        private val type: TextView? = itemView.findViewById(R.id.type)
        private val pinned: ImageView? = itemView.findViewById(R.id.pinned)

        fun onBind(item: JsonObject) {
            if (item.text == null) {
                return
            }
            setImage(item)
            setType(item)
            setTitle(item)
            message?.text = Html.fromHtml(item.text?.replace("<p>", "")?.replace("</p>", ""))
            unread?.setUnreadCount(item.unreadCount ?: 0)
            pinned?.visibility = if ((item.unreadCount ?:0) <= 0 && (item.teamPin ?:0F) > 0F) View.VISIBLE else View.INVISIBLE

            item.itemDate?.let {
                date?.setRelativeTimeSpan(it)
            }

            item.topPosterAvatars?.map { it.asString }?.let {
                avatars?.setAvatars(imageLoader, it, item.posterCount ?: 0)
            }
            setClickAction(item)

        }


        private fun setImage(item: JsonObject) {
            item.smallPhotoOrAvatar?.let { imageLoader.getImageUrl(it) }?.let {
                when (item.itemType) {
                    TeambrellaModel.FEED_ITEM_TEAMMATE,
                    TeambrellaModel.FEED_ITEM_TEAM_CHAT -> {
                        icon?.setAvatar(it)
                    }
                    else -> {
                        icon?.setImage(it, R.dimen.rounded_corners_2dp)
                    }
                }
            }
        }

        private fun setType(item: JsonObject) {
            date?.visibility = View.VISIBLE
            avatars?.visibility = View.VISIBLE
            unread?.visibility = View.VISIBLE
            type?.visibility = View.VISIBLE
            message?.maxLines = 2
            pane?.setBackgroundColor(Color.WHITE)

            when (item.itemType) {
                TeambrellaModel.FEED_ITEM_CLAIM -> {
                    type?.setText(R.string.claim)
                    type?.setRightCompoundDrawable(R.drawable.ic_claim)
                }
                
                TeambrellaModel.FEED_ITEM_TEAM_CHAT -> {
                    type?.setRightCompoundDrawable(R.drawable.ic_discussion)
                    type?.setText(R.string.discussion)
                }
    
                TeambrellaModel.FEED_ITEM_FUND_WALLET -> {
                    date?.visibility = View.GONE
                    avatars?.visibility = View.GONE
                    unread?.visibility = View.GONE
                    type?.visibility = View.GONE
                    message?.maxLines = 4
                    pane?.setBackgroundColor(itemView.context.getResources().getColor(R.color.veryLightBlueTwo))
                }
                
                else -> {
                    //type?.setRightCompoundDrawable(R.drawable.ic_application)
                    //type?.setText(R.string.application)
                    type?.setCompoundDrawablesWithIntrinsicBounds(null, null, null, null);
                    type?.setText("")
                }
            }
        }

        private fun setTitle(item: JsonObject) {
            when (item.itemType) {
                TeambrellaModel.FEED_ITEM_TEAM_CHAT -> title?.text = item.chatTitle
                TeambrellaModel.FEED_ITEM_TEAMMATE -> title?.text = item.itemUserName
                TeambrellaModel.FEED_ITEM_FUND_WALLET -> title?.text = item.chatTitle
                else -> title?.text = item.modelOrName
            }
        }

        private fun setClickAction(item: JsonObject) {
            itemView.setOnClickListener {
                when (item.itemType) {
                    TeambrellaModel.FEED_ITEM_CLAIM -> {
                        startActivity(ChatActivity.getClaimChat(
                                itemView.context,
                                teamId,
                                item.itemIdInt ?: 0,
                                item.modelOrName,
                                item.smallPhotoOrAvatar,
                                item.topicId,
                                dataHost.teamAccessLevel,
                                item.itemDate))
                    }

                    TeambrellaModel.FEED_ITEM_TEAM_CHAT -> {
                        startActivity(ChatActivity.getFeedChat(
                                itemView.context,
                                item.chatTitle,
                                item.topicId,
                                teamId,
                                dataHost.teamAccessLevel))
                    }
    
                    TeambrellaModel.FEED_ITEM_FUND_WALLET -> {
                        dataHost.showWallet()
                    }
                    
                    else -> {
                        startActivity(ChatActivity.getTeammateChat(
                                itemView.context, teamId,
                                item.itemUserId,
                                item.itemUserName,
                                item.smallPhotoOrAvatar,
                                item.topicId,
                                dataHost.teamAccessLevel))
                    }
                }
            }

        }
    }

    private inner class FeedHeader internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {
        init {
            itemView.findViewById<View>(R.id.start_new_discussion).setOnClickListener {
                dataHost.startNewDiscussion()
            }
        }
    }
}